lib/vrl/stdlib/src/parse_csv.rs

```diff
@@ -0,0 +1,130 @@
+use bytes::Buf;
+use csv::ReaderBuilder;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCsv;
+
+impl Function for ParseCsv {
+    fn identifier(&self) -> &'static str {
+        "parse_csv"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse a single CSV formatted row",
+            source: r#"parse_csv!("foo,bar,"foo "", bar"")"#,
+            result: Ok(r#"["foo", "bar", "foo ", bar"]"#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let delimiter = arguments.optional("delimiter").unwrap_or(expr!(","));
+        Ok(Box::new(ParseCsvFn { value, delimiter }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "delimiter",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCsvFn {
+    value: Box<dyn Expression>,
+    delimiter: Box<dyn Expression>,
+}
+
+impl Expression for ParseCsvFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let csv_string = self.value.resolve(ctx)?.try_bytes()?;
+        let delimiter = self.delimiter.resolve(ctx)?.try_bytes()?;
+        if delimiter.len() != 1 {
+            return Err("delimiter must be a single character".into());
+        }
+        let delimiter = delimiter[0];
+
+        let reader = ReaderBuilder::new()
+            .has_headers(false)
+            .delimiter(delimiter)
+            .from_reader(csv_string.bytes());
+
+        let parsed_record = reader.into_records().next();
+        match parsed_record {
+            Some(_) => {
+                let result = parsed_record
+                    .unwrap()
```

Comment:
Should we be `unwrap()`ing here? Under what conditions could it fail?

Looking at:

https://docs.rs/csv/1.1.6/csv/struct.Reader.html#method.into_records

It seems like we should be bubbling up any errors from here.

Reply:
Indeed, though unlikely but seems that we can have some errors here, e.g. when we have invalid utf8 string. Hope that corrected error handling properly here.

Reply:
refactored quite a bit as @JeanMertz suggested(please see the comment below).